spec/lib/msf/ui/console/command_dispatcher/exploit_spec.rb

```diff
@@ -1,30 +1,343 @@
+    instance
   end
 
-  describe "#cmd_rcheck" do
+  # Intercepts calls to Rex::Job objects, and ensures that async rex jobs are immediately run inline instead of having
+  # their execution deferred until later. This ensures that Jobs deterministically complete during a test run.
+  def run_rex_jobs_inline!
+    allow_any_instance_of(Rex::Job).to receive(:start).and_wrap_original do |original_method, original_async_value|
+      original_receiver = original_method.receiver
+      mod = original_receiver.ctx[0]
+      mod.print_status("Running rex job #{original_receiver.jid} inline")
+      expect(original_async_value).to be(true)
+      new_async_value = false
+      original_method.call(new_async_value)
+    end
   end
 
-  describe "#cmd_rexploit" do
+  def set_default_payload(mod)
+    mod.datastore['PAYLOAD'] = 'generic/no_session_payload'
+    mod.datastore['LHOST'] = '127.0.0.1'
+    mod.datastore['RHOSTS'] = '192.0.2.1'
   end
 
-  describe "#cmd_reload" do
+  before do
+    run_rex_jobs_inline!
+
+    allow(driver).to receive(:input).and_return(driver_input)
+    allow(driver).to receive(:output).and_return(driver_output)
+    current_mod.init_ui(driver_input, driver_output)
+    allow(subject).to receive(:mod).and_return(current_mod)
+
+    framework.modules.add_module_path(File.join(FILE_FIXTURES_PATH, 'modules'))
   end
 
-  describe "#cmd_run" do
+  describe '#cmd_check' do
+    context 'when running an exploit module' do
+      let(:current_mod) { exploit_mod }
+
+      it 'reports missing RHOST values' do
+        allow(current_mod).to receive(:run).and_call_original
+        current_mod.datastore['RHOSTS'] = ''
+        subject.cmd_check
+        expected_output = [
+          'Check failed: Msf::OptionValidateError One or more options failed to validate: RHOSTS.'
+        ]
+
+        expect(@combined_output).to match_array(expected_output)
+        expect(subject.mod).not_to have_received(:run)
+      end
+
+      it 'runs a single RHOST value' do
+        current_mod.datastore['RHOSTS'] = '192.0.2.1'
+        subject.cmd_check
+        expected_output = [
+          'Checking for target 192.0.2.1:3000 with normalized datastore value 3.5',
+          'Cleanup for target 192.0.2.1:3000',
+          '192.0.2.1:3000 - Check failed: The state could not be determined.'
+        ]
+
+        expect(@combined_output).to match_array(expected_output)
+      end
+
+      it 'runs multiple RHOST values' do
+        current_mod.datastore['RHOSTS'] = '192.0.2.1 192.0.2.2'
+        subject.cmd_check
+        expected_output = [
+          'Checking for target 192.0.2.1:3000 with normalized datastore value 3.5',
+          '192.0.2.1:3000 - Check failed: The state could not be determined.',
+          'Checking for target 192.0.2.2:3000 with normalized datastore value 3.5',
+          '192.0.2.2:3000 - Check failed: The state could not be determined.',
+          'Cleanup for target 192.0.2.1 192.0.2.2:3000'
```

Comment:
I haven't looked into why the cmd_check functionality cleans up modules twice yet, but it seemed like a bug to me on my rhosts uri support branch too